onnxruntime/core/providers/xnnpack/nn/unpool.cc

```diff
@@ -0,0 +1,266 @@
+
+// to sanity check output shape
+#include "core/framework/tensorprotoutils.h"
+#include "gsl/gsl-lite.hpp"
+
+namespace onnxruntime {
+namespace xnnpack {
+namespace {
+
+TensorShapeVector InferOutputSizeForUnPool(const PoolAttributes& pool_attrs,
+                                           const TensorShape& input_shape) {
+  // Calculate output tensor shape from attributes
+  TensorShapeVector inferred_output_dims(input_shape.NumDimensions());
+
+  // Copy batch and channel dims
+  inferred_output_dims[0] = input_shape[0];
+  inferred_output_dims[3] = input_shape[3];
+
+  // For feature dims calculate reversing the formula used for MaxPool
+  for (size_t dim = 0; dim < pool_attrs.kernel_shape.size(); ++dim) {
+    inferred_output_dims[dim + 1] =
+        (input_shape[dim + 1] - 1) * pool_attrs.strides[dim] -
+        (pool_attrs.pads[dim] + pool_attrs.pads[pool_attrs.kernel_shape.size() + dim]) +
+        pool_attrs.kernel_shape[dim];
+  }
+
+  return inferred_output_dims;
+}
+}  // namespace
+
+// MaxUnpool doesn't have any quantization params
+bool MaxUnpool::IsOnnxNodeSupported(const NodeUnit& node_unit,
+                                    const GraphViewer& graph_viewer) {
+  ORT_UNUSED_PARAMETER(graph_viewer);
+  bool supported = false;
+  const onnxruntime::Node& node = node_unit.GetNode();
+  // use do {} while(false) so it's easier to set a breakpoint on the return
+  do {
+    // MaxUnpool has 2-3 inputs.
+    auto input_defs = node.InputDefs();
+
+    if (input_defs.size() == 3) {
+      const auto* output_shape = graph_viewer.GetConstantInitializer(input_defs[2]->Name(), true);
+      if (output_shape == nullptr || output_shape->dims_size() != 1 || output_shape->dims(0) != 4) {
+        break;
+      }
+    }
+    const auto& x_arg = *input_defs[0];
+
+    const auto* x_type = x_arg.TypeAsProto();
+    if (x_type == nullptr ||
+        (x_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT)) {
+      break;
+    }
+
+    // we only support 2D (4 dims with batch and channel)
+    const auto* x_shape = x_arg.Shape();
+    if (!x_shape || x_shape->dim_size() != 4) {
+      break;
+    }
+
+    // require C, H, W to be known so we can construct the xnnpack kernel prior to Compute
+    if (!x_shape->dim(1).has_dim_value() ||
+        !x_shape->dim(2).has_dim_value() ||
+        !x_shape->dim(3).has_dim_value()) {
+      break;
+    }
+
+    ProtoHelperNodeContext nc(node);
+    OpNodeProtoHelper info(&nc);
+    PoolAttributes pool_attrs(info, "MaxUnpool", node.SinceVersion());
+
+    if (!IsPaddingTypeSupported(pool_attrs.auto_pad)) {
+      break;
+    }
+
+    supported = true;
+  } while (false);
+
+  return supported;
+}
+
+MaxUnpool::MaxUnpool(const OpKernelInfo& info)
+    : XnnpackKernel(info),
+      pool_attrs_{info, "MaxUnpool", info.node().SinceVersion()} {
+  num_inputs_ = OpKernel::Node().InputDefs().size();
+  info.GetAttrOrDefault<int64_t>("mode", &is_indice_produced_by_xnnpack_, 0);
```

Comment:
MaxUnpool doesn't have a 'mode' attribute. Can you explain where that comes from and how it changes what the implementation should do?

Reply:
Yes, It does not have 'mode'  in MaxUnpool.
The problem is that xnnpack_maxunpool should be feed in a input tensor and a indice-tensor (uint32_t). besides,  indice-tensor in xnnpack have totally different definition, it cost too much if transform from onnx-definition to xnnpack definition. please Refer https://github.com/google/XNNPACK/blob/04f03be0b8485fd0e30e4b6d402da198d4463a40/test/unpooling-operator-tester.h#L301

So we need a attribute `mode` to distinguish `xnn_create_unpooling2d_nhwc_x32` and nhwc_maxunpool.
Or we remove all xnn_create_unpooling2d_nhwc_x32 related code.

Reply:
All removed.